### PR TITLE
[Core] Fix triangle division when one of the nodes has exactly zero distance

### PR DESCRIPTION
**Pull request by jcotela**
**📝 Description**
In the current master DivideTriangle3D3 throws an error if the nodal distance is exactly zero, even though it is actually doing the right thing. This is due to the part where it counts positive and negative nodes, which always counts 0 as positive and reports a change of sign for the negative side triangle.

I assume the same is true for the 2D version, since the proposed change actually involves triangle_2d_3.cpp, but I have not used it.

This PR addresses the issue and introduces a test illustrating the expected behavior.

**🆕 Changelog**
- Fixed issue in DivideTriangle3D3 when one of the nodes has exactly zero distance.

### CODE REVIEW

File: kratos/utilities/divide_triangle_2d_3.cpp
Comment (loumalouomega) [inline kratos/utilities/divide_triangle_2d_3.cpp]: KRATOS_ERROR_IF

**Review by rubenzorrilla — COMMENTED** [review]
This is a very delicate corner case, which I admittedly haven't had the time to carefully look into. Though the splitting algorithm may work, I think we should test how this splitting pattern behaves when used in combination with the modified shape functions calculation.
Hence, I'd suggest to add the very same test set in there as well. Besides, I think this is also happening in the tetrahedron.

File: kratos/tests/cpp_tests/utilities/test_divide_triangle_3d_3.cpp
Comment (rubenzorrilla) [inline kratos/tests/cpp_tests/utilities/test_divide_triangle_3d_3.cpp:295]: I think this test must be added in the `test_divide_triangle_2d_3.cpp` instead of this one.

**Comment by jcotela** [comment]
I am not using this in the context of enriched shape functions, but only as part of a tool to subdivide triangles in a poor quality mesh that comes from an STL, so what you see here is just a reformulation of a weird case I found as a test and its proposed fix... I guess that in the context of enrichment this may be a non-issue, since otherwise you would have noticed by now (I would think the nudging to prevent bad cuts prevents it?).

All this is to say that I will need some guidance if extended testing is needed, since I am not that sure of what to look for in the context of enrichment.

File: kratos/tests/cpp_tests/utilities/test_divide_triangle_3d_3.cpp
Comment (jcotela) [reply kratos/tests/cpp_tests/utilities/test_divide_triangle_3d_3.cpp:295]: I would add both (it would not be the first time in Kratos that something that works in XY fails in YZ)

**Comment by rubenzorrilla** [comment]
> I am not using this in the context of enriched shape functions, but only as part of a tool to subdivide triangles in a poor quality mesh that comes from an STL, so what you see here is just a reformulation of a weird case I found as a test and its proposed fix... I guess that in the context of enrichment this may be a non-issue, since otherwise you would have noticed by now (I would think the nudging to prevent bad cuts prevents it?).
> 
> All this is to say that I will need some guidance if extended testing is needed, since I am not that sure of what to look for in the context of enrichment.

My point is that even though you aim at using the splitting utilities out of the context of the enrichment, the behavior of the former is crucial for the latter, so I want to be sure that changing what is considered OK to be split doesn't affect the enrichment utils. I don't think we can rely on the ill-conditioned cuts prevention as this is one of the possible alternatives in literature to make the formulation stable.  In this regard, I note that a zero distance value might not imply a "zero"-cut.

Having said this, my suggestion would be to mimic one of the tests in [here](https://github.com/KratosMultiphysics/Kratos/blob/8067b99315faa2a828f277b7febdcbcc246e1729/kratos/tests/cpp_tests/modified_shape_functions/test_triangle_2d_3_modified_shape_functions.cpp) but with the triangle and distance pattern you're using in the test in this PR.

**Comment by jcotela** [comment]
I agree that it is important to keep the enriched element working, but the situation that this PR attempts to fix results in a thrown exception, so it is quite evident when it happens in practice. If you are not seeing simulations fail at random as soon as an element is cut in this way, it seems to me that it is not a common situation.

For the 2D triangle test, I reproduced the 3D configuration for the XY plane in [98985b3](https://github.com/KratosMultiphysics/Kratos/pull/11299/commits/98985b347d2fc3a0553d4025ebfbd5624ec84fee). I think that the only thing missing would be a test using the relevant enriched element for a minimal simulation (or at least to get its linear system contribution), but I am not sure how to do it.

**Comment by RiccardoRossi** [comment]
The enrichment will definitely fail in this context, but as @jcotela suggests it should not be happening as of now since if it was an error would be thrown.

the right fix would be to have the error thrown in the calculation of the enrichment funciton rather than in the splitting utility, so i would agree with   @jcotela about merging this.

**Review by RiccardoRossi — DISMISSED** [review]
to me the argument of @jcotela is valid so i am approving, however in case it is not urgent and u can wait for further feedback from  @rubenzorrilla please do

File: kratos/tests/cpp_tests/utilities/test_divide_triangle_2d_3.cpp
Comment (loumalouomega) [inline kratos/tests/cpp_tests/utilities/test_divide_triangle_2d_3.cpp]: This tests have tabs!

**Comment by RiccardoRossi** [comment]
@jcotela why not merging?

**Comment by rubenzorrilla** [comment]
> @jcotela why not merging?

Well, I'd like to have the modified shape functions test I suggested above before merging this.

**Comment by jcotela** [comment]
@rubenzorrilla which element should the tests use? Simplest thing would be to point me to a test using an element that requires this utility and I can replicate it with the configuration that triggers this error.

**Comment by rubenzorrilla** [comment]
> > I am not using this in the context of enriched shape functions, but only as part of a tool to subdivide triangles in a poor quality mesh that comes from an STL, so what you see here is just a reformulation of a weird case I found as a test and its proposed fix... I guess that in the context of enrichment this may be a non-issue, since otherwise you would have noticed by now (I would think the nudging to prevent bad cuts prevents it?).
> > All this is to say that I will need some guidance if extended testing is needed, since I am not that sure of what to look for in the context of enrichment.
> 
> My point is that even though you aim at using the splitting utilities out of the context of the enrichment, the behavior of the former is crucial for the latter, so I want to be sure that changing what is considered OK to be split doesn't affect the enrichment utils. I don't think we can rely on the ill-conditioned cuts prevention as this is one of the possible alternatives in literature to make the formulation stable. In this regard, I note that a zero distance value might not imply a "zero"-cut.
> 
> Having said this, my suggestion would be to mimic one of the tests in [here](https://github.com/KratosMultiphysics/Kratos/blob/8067b99315faa2a828f277b7febdcbcc246e1729/kratos/tests/cpp_tests/modified_shape_functions/test_triangle_2d_3_modified_shape_functions.cpp) but with the triangle and distance pattern you're using in the test in this PR.

@jcotela I attached a link in my previous post (see the one in "here"). My idea is to test the modified shape functions but with the "new" splitting pattern (i.e. the zero distance).

**Comment by jcotela** [comment]
@rubenzorrilla I have added a test based on the one you provided. To make it pass I had to fix the calculation of the interface too, but I think it is working now.

I also did a second small modification to make sure that triangles with two zero nodes are not incorrectly cut (and added a test). I can do an extra commit with whitespace since changes as @loumalouomega noticed that the test files use tabs, but otherwise I think this is good to go.

File: kratos/utilities/divide_triangle_2d_3.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_triangle_2d_3.cpp]:
```suggestion
                // Determine if the subdivision is whether in the negative or the positive side
                // Note that zero distance nodes are also identified and stored in here 
```

File: kratos/utilities/divide_triangle_2d_3.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_triangle_2d_3.cpp]:
```suggestion
                    // If the indexed keys is larger or equal to the number of nodes means that they are the auxiliary interface points
                    // For the zero distance case, the corresponding node is considered as part of the interface
```

File: kratos/utilities/divide_triangle_2d_3.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_triangle_2d_3.cpp]:
```suggestion
                    // If the indexed keys is larger or equal to the number of nodes means that they are the auxiliary interface points
                    // For the zero distance case, the corresponding node is considered as part of the interface
```

**Review by rubenzorrilla — COMMENTED** [review]
Now it looks good to me. I'll approve if you agree with my documentation suggestions.

However, right now the behavior of the triangle and tetrahedron splitting is inconsistent. I think we should update it accordingly.

**Comment by jcotela** [comment]
No problem regarding the comments (I don't think I touched those, though?)

Regarding the tets, I fully agree. I believe that the same approach used here should be applicable, but it is something that should be done and tested carefully

**Comment by rubenzorrilla** [comment]
> No problem regarding the comments (I don't think I touched those, though?)
> 
> Regarding the tets, I fully agree. I believe that the same approach used here should be applicable, but it is something that should be done and tested carefully

I think that for a zero node distance we'll get 2 lets say positive subtets and a negative one. For two zero node distance case I think we'll get one positive subtet and a negative one. For three or more zero distance nodes there won't be splitting.

Having said this, shall we do it in this same PR or in a future one?

**Comment by jcotela** [comment]
@rubenzorrilla I've gone ahead and applied the same changes to the tetrahedra divider. I think it works (I added a test) but I won't be using it, so I'd encourage you to make sure it is ok.

File: kratos/utilities/divide_tetrahedra_3d_4.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_tetrahedra_3d_4.cpp]:
```suggestion
                    // If the indexed keys is larger or equal to the number of nodes means that they are the auxiliary interface points
                    // For the zero distance case, the corresponding node is considered as part of the interface
```

File: kratos/utilities/divide_tetrahedra_3d_4.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_tetrahedra_3d_4.cpp]:
```suggestion
                    // If the indexed keys is larger or equal to the number of nodes means that they are the auxiliary interface points
                    // For the zero distance case, the corresponding node is considered as part of the interface
```

File: kratos/utilities/divide_tetrahedra_3d_4.h
Comment (rubenzorrilla) [inline kratos/utilities/divide_tetrahedra_3d_4.h:189]: This is admittedly a bit picky, but I'd add this method to the triangle divider as well (it makes the code more self-comprehensive).

File: kratos/utilities/divide_tetrahedra_3d_4.cpp
Comment (rubenzorrilla) [inline kratos/utilities/divide_tetrahedra_3d_4.cpp]:
```suggestion
                // Determine if the subdivision is whether in the negative or the positive side
                // Note that zero distance nodes are also identified and stored in here 
```

**Review by rubenzorrilla — COMMENTED** [review]
I think that what you did is correct. My unique suggestion would be to add a test for the zero node tetrahedra modified shape functions calculator as you did for the triangles one.

**Comment by jcotela** [comment]
I think this is done now

**Review by rubenzorrilla — APPROVED** [review]
Great. Thanks for the patience.

**Pull request merged**
